File: sharov/Plastomer/Unit_About.pas
unit Unit_About;

interface

uses Windows, SysUtils, Classes, Graphics, Forms, Controls, StdCtrls,
  Buttons, ExtCtrls;

type
  TfrmAbout = class(TForm)
    OKButton: TButton;
    Panel1: TPanel;
    Image1: TImage;
    ProductName: TLabel;
    Version: TLabel;
    Comments: TLabel;
    Label1: TLabel;
    Label3: TLabel;
    PhysMem: TLabel;
    FreeRes: TLabel;
    CpuLabel: TLabel;
    Image2: TImage;
    Label2: TLabel;
    Label5: TLabel;
    Label6: TLabel;
    Label4: TLabel;
    procedure Label2Click(Sender: TObject);
    procedure OKButtonClick(Sender: TObject);
    procedure FormActivate(Sender: TObject);
  private
    { Private declarations }
  public


    { Public declarations }
  end;

var
  frmAbout: TfrmAbout;

implementation

uses ShellApi, mpiDeclaration;

{$R *.DFM}

procedure TfrmAbout.Label2Click(Sender: TObject);
begin
  ShellExecute(Handle,'open', PChar(TLabel(Sender).Caption),nil,nil,SW_SHOW);
end;

procedure TfrmAbout.OKButtonClick(Sender: TObject);
begin
  Close;
end;

function GetCPUSpeed: Double;
const
  DelayTime = 200;
var
  TimerHi, TimerLo: Cardinal;
  PriorityClass, Priority: Integer;
begin
  PriorityClass := GetPriorityClass(GetCurrentProcess);
  Priority := GetThreadPriority(GetCurrentThread);
  SetPriorityClass(GetCurrentProcess, REALTIME_PRIORITY_CLASS);
  SetThreadPriority(GetCurrentThread, THREAD_PRIORITY_TIME_CRITICAL);
  Sleep(10);
  asm
    dw 310Fh // rdtsc
    mov TimerLo, eax
    mov TimerHi, edx
  end;
  Sleep(DelayTime);
  asm
    dw 310Fh // rdtsc
    sub eax, TimerLo
    sbb edx, TimerHi
    mov TimerLo, eax
    mov TimerHi, edx
  end;
  SetThreadPriority(GetCurrentThread, Priority);
  SetPriorityClass(GetCurrentProcess, PriorityClass);
  Result := TimerLo / (1000.0 * DelayTime);
end;

procedure TfrmAbout.FormActivate(Sender: TObject);
var
  MS: TMemoryStatus;
begin
  ProductName.Caption := ProjectCaption;

  GlobalMemoryStatus(MS);

  PhysMem.Caption  := 'Всего физической памяти : '      + FormatFloat('#,###" KB"', MS.dwTotalPhys / 1024);
  FreeRes.Caption  := 'Используется в данный момент : ' + Format('%d %%', [MS.dwMemoryLoad]);
  CpuLabel.Caption := 'Частота процессора : ';
  
  Application.ProcessMessages;
  CpuLabel.Caption := CpuLabel.Caption + FloatToStr(Round(GetCPUSpeed*10)/10)+' МГц';
end;

end.

File: kvant/Far_2/Far_2/Project_Old_Old/Unit_About.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit_About.h"
#include "mpiDeclaration.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmAbout *frmAbout;
//---------------------------------------------------------------------------
__fastcall TfrmAbout::TfrmAbout(TComponent* Owner)
        : TForm(Owner)
{
//   Image1->Transparent = true;
//   Image1->Picture->Bitmap->TransparentColor = clWhite;
}
//---------------------------------------------------------------------------
void __fastcall TfrmAbout::FormShow(TObject *Sender)
{
  Caption = ProjectCaption + "О программе ...";

  Panel2->Top = 2;
}

//---------------------------------------------------------------------------
void __fastcall TfrmAbout::FormClick(TObject *Sender)
{
   CountDblClick = 0;
   if (!Timer1->Enabled) Panel2->Top = 2;
}
//---------------------------------------------------------------------------
void __fastcall TfrmAbout::Image1DblClick(TObject *Sender)
{
  CountDblClick ++;
  if (CountDblClick == 3) {
    Timer1->Enabled = true;
    CountDblClick = 0;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmAbout::Timer1Timer(TObject *Sender)
{
  Panel2->Top --;
  if (Panel2->Top + Panel2->Height == Panel1->Height) {
     Timer1->Enabled = false;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmAbout::Label3Click(TObject *Sender)
{
   try    {
      ShellExecute(Handle, "open", ((TLabel*)Sender)->Caption.c_str(), NULL, NULL, SW_SHOW);
   } catch  ( ... ) {
      MessageDlg("Ошибка открытия браузера", mtError, TMsgDlgButtons() << mbOK, 0);
      return;
   }
}

//---------------------------------------------------------------------------
double __fastcall TfrmAbout::GetCPUSpeed()
{
  int       DelayTime = 200;
  Cardinal  TimerHi;
  Cardinal  TimerLo;

  int PriorityClass = GetPriorityClass(GetCurrentProcess);
  int Priority      = GetThreadPriority(GetCurrentThread);

  SetPriorityClass(GetCurrentProcess, REALTIME_PRIORITY_CLASS);
  SetThreadPriority(GetCurrentThread, THREAD_PRIORITY_TIME_CRITICAL);

  Sleep(10);

  asm {
    dw  0x310F // rdtsc
    mov TimerLo, eax
    mov TimerHi, EDX
  }

  Sleep(DelayTime);

  asm {
    dw  0x310F // rdtsc
    sub eax, TimerLo
    sbb edx, TimerHi
    mov TimerLo, eax
    mov TimerHi, edx
  }

  SetThreadPriority(GetCurrentThread, Priority);
  SetPriorityClass(GetCurrentProcess, PriorityClass);

  return TimerLo / (1000.0 * DelayTime);
}

//---------------------------------------------------------------------------
void __fastcall TfrmAbout::FormActivate(TObject *Sender)
{
 MEMORYSTATUS MS;

  GlobalMemoryStatus(&MS);

  PhysMem->Caption  = "Всего физической памяти : "      + FormatFloat("#,###' KB'", MS.dwTotalPhys / 1024.0);
  FreeRes->Caption  = "Используется в данный момент : " + IntToStr(MS.dwMemoryLoad) + " %";
  CpuLabel->Caption = "Частота процессора : ";

  CpuLabel->Caption = CpuLabel->Caption + FloatToStr(int((GetCPUSpeed() * 0.9989544010)*10.0)/10.0) + " МГц";

  DecimalSeparator = '.';
}
//---------------------------------------------------------------------------
void __fastcall TfrmAbout::BitBtn1Click(TObject *Sender)
{
   CountDblClick = 0;
   Timer1->Enabled = false;
}
//---------------------------------------------------------------------------

